### PR TITLE
Store first and last search time and add filter options

### PR DESCRIPTION
**Pull request by craigjmidwinter**
Not sure if this is what you had in mind for how to approach the timestamp filtering, but I thought I would take a stab at it.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -40,6 +46,16 @@ async function findOnOtherSites(info, hashesToExclude) {
...
 }
 
+async function updateSearchTimestamps(infoHash) {
+	const cacheKey = CACHE_PREFIX_TIMESTAMPS + "|" + infoHash;
+	const existingTimestamps = get(cacheKey);
+	const firstSearched = existingTimestamps
+		? existingTimestamps.firstSeen
+		: new Date();
+	const lastSearched = new Date();
+	save(cacheKey, { firstSearched, lastSearched });
```
Comment (mmgoodnow) [inline src/index.js:56]: I think `existingTimestamps.firstSeen` should be `existingTimestamps.firstSearched`

File: src/preFilter.js [hunk 2]
```diff
@@ -59,4 +60,29 @@ function filterDupes(metafiles) {
...
+	if (tsFirstSearched < tsOldCutoff) {
+		return false;
+	}
+
+	const tsRecentCutoff =
+		new Date().getTime() - excludeRecentSearch * 60 * 1000;
+	const tsLastSearched = new Date(timestampData.lastSearched).getTime();
+	if (tsLastSearched > tsRecentCutoff) {
+		return false;
+	}
```
Comment (mmgoodnow) [inline src/preFilter.js:83]:
```suggestion
	if (
		excludeRecentSearch && 
		lastSearched > nMinutesAgo(excludeRecentSearch)
	) {
		logReason(...);
		return false;
	}
```

File: src/cache.js [hunk 3]
```diff
@@ -1,35 +1,56 @@
...
 	}
 }
 
+function migrateV1Cache(fileCache) {
+	const newCache = {};
+	fileCache.forEach((cacheEntry) => {
+		newCache[cacheEntry] = true;
+	});
+	return newCache;
+}
```
Comment (mmgoodnow) [inline src/cache.js:27]: Good idea to migrate the old cache.

File: src/cache.js [hunk 4]
```diff
@@ -1,35 +1,56 @@
...
+	try {
+		return cache[key];
+	} catch (e) {
+		return null;
+	}
+}
+
 function includes(thing) {
-	return cache.includes(thing);
+	return Object.keys(cache).includes(thing);
```
Comment (mmgoodnow) [inline src/cache.js:49]:
```suggestion
	return cache.hasOwnProperty(thing);
```

**Review by mmgoodnow — CHANGES_REQUESTED** [review]
Wow, thanks so much! This is my first time receiving a PR and I really appreciate the work you've done :)

I've pointed out a few things I'd like to change before merging this into master. Would you mind if I hop in and make those changes? You could also do it if you'd like.

**Comment by craigjmidwinter** [comment]
> Wow, thanks so much! This is my first time receiving a PR and I really appreciate the work you've done :)
> 
> I've pointed out a few things I'd like to change before merging this into master. Would you mind if I hop in and make those changes? You could also do it if you'd like.

Yeah all those changes make sense to me! I can try make them either later tonight or tomorrow evening, or if you have time and are wanting to jump in and make them to get it out, feel free!

**Comment by mmgoodnow** [comment]
cool, go for it!

**Comment by craigjmidwinter** [comment]
Updated as per your notes! The one exception being using the prototype field of hasOwnProperty since eslint forbids it.

File: src/index.js [hunk 5]
```diff
@@ -40,6 +46,16 @@ async function findOnOtherSites(info, hashesToExclude) {
 	return successful.length;
 }
 
+async function updateSearchTimestamps(infoHash) {
+	const cacheKey = CACHE_PREFIX_TIMESTAMPS + infoHash;
+	const existingTimestamps = get(cacheKey);
+	const firstSearched = existingTimestamps
+		? existingTimestamps.firstSearched
+		: new Date().getTime();
```
Comment (mmgoodnow) [inline src/index.js:54]: super nit: can we change these to Date.now()?

**Review by mmgoodnow — COMMENTED** [review]
It's coming along nicely! Just a couple small things left.

**Comment by craigjmidwinter** [comment]
Updated!

**Review by mmgoodnow — APPROVED** [review]
Great work, thanks for the contribution! I'll test this out this evening, and then merge the PR and update the README with the new options.

**Pull request merged**